pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1151 @@
+                if removed {
+                    timeline.metrics.resident_physical_size_gauge.sub(file_size);
+                }
+                if let Some(remote_client) = timeline.remote_client.as_ref() {
+                    let res = remote_client.schedule_layer_file_deletion(&[file_name]);
+
+                    if let Err(e) = res {
+                        // test_timeline_deletion_with_files_stuck_in_upload_queue is good at
+                        // demonstrating this deadlock (without spawn_blocking): stop will drop
+                        // queued items, which will have ResidentLayer's, and those drops would try
+                        // to re-entrantly lock the RemoteTimelineClient inner state.
+                        if !timeline.is_active() {
+                            tracing::info!("scheduling deletion on drop failed: {e:#}");
+                        } else {
+                            tracing::warn!("scheduling deletion on drop failed: {e:#}");
+                        }
+                    }
+                }
+            } else {
+                // no need to nag that timeline is gone: under normal situation on
+                // task_mgr::remove_tenant_from_memory the timeline is gone before we get dropped.
+            }
+        });
+    }
+}
+
+impl LayerInner {
+    fn new(
+        conf: &'static PageServerConf,
+        timeline: &Arc<Timeline>,
+        access_stats: LayerAccessStats,
+        desc: PersistentLayerDesc,
+        downloaded: Option<Arc<DownloadedLayer>>,
+    ) -> Self {
+        let path = conf
+            .timeline_path(&timeline.tenant_id, &timeline.timeline_id)
+            .join(desc.filename().to_string());
+
+        LayerInner {
+            conf,
+            path,
+            desc,
+            timeline: Arc::downgrade(timeline),
+            have_remote_client: timeline.remote_client.is_some(),
+            access_stats,
+            wanted_garbage_collected: AtomicBool::new(false),
+            wanted_evicted: AtomicBool::new(false),
+            inner: if let Some(inner) = downloaded {
+                heavier_once_cell::OnceCell::new(ResidentOrWantedEvicted::Resident(inner))
+            } else {
+                heavier_once_cell::OnceCell::default()
+            },
+            version: AtomicUsize::new(0),
+            status: tokio::sync::broadcast::channel(1).0,
+            consecutive_failures: AtomicUsize::new(0),
+        }
+    }
+
+    fn garbage_collect_on_drop(&self) {
+        self.wanted_garbage_collected.store(true, Ordering::Release);
+    }
+
+    pub(crate) async fn evict_and_wait(
+        &self,
+        _: &RemoteTimelineClient,
+    ) -> Result<(), EvictionError> {
+        use tokio::sync::broadcast::error::RecvError;
+
+        assert!(self.have_remote_client);
+
+        self.wanted_evicted.store(true, Ordering::Release);
+
+        let mut rx = self.status.subscribe();
+
+        if self.get().is_none() {
```

Comment:
I was in fact thinking of using a completion within DownloadedLayer so it would be more stricly about *one* round of eviction, but I think your suggestion to reorder the two is good regardless.

Reply:
Reordered in c659d0f21